File: src/main/java/ro/sopa/statistifier/job/JobRegistry.java
package ro.sopa.statistifier.job;

import org.springframework.web.context.annotation.ApplicationScope;
import ro.sopa.statistifier.UserImportException;

import java.util.ArrayList;
import java.util.List;

@ApplicationScope
public class JobRegistry {

    private List<RunningJob> runningJobs = new ArrayList<>();

    public RunningJob register(JobDescriptor descriptor) {
        RunningJob rj = findRunningJob(descriptor);

        if(rj != null) {
            throw new RuntimeException("Job already running for " + descriptor.toString() + " and is at " + rj.getStep() + "/" + rj.getTotal());
        }

        rj = new RunningJob(descriptor);

        runningJobs.add(rj);

        return rj;
    }

    public void setTotalStep(JobDescriptor descriptor, int total) {
        RunningJob runningJob = findRunningJob(descriptor);

        if(runningJob == null) {
            throw new RuntimeException("Invalid job: " + descriptor.toString());
        }

        runningJob.setTotal(total);
    }

    public void setStep(JobDescriptor descriptor, int step) {
        RunningJob runningJob = findRunningJob(descriptor);

        if(runningJob == null) {
            throw new RuntimeException("Invalid job: " + descriptor.toString());
        }

        runningJob.setStep(step);
    }

    public void finish(JobDescriptor descriptor) {
        RunningJob runningJob = findRunningJob(descriptor);

        if(runningJob == null) {
            throw new RuntimeException("Invalid job: " + descriptor.toString());
        }

        runningJobs.remove(runningJob);
    }

    public RunningJob findRunningJob(JobDescriptor descriptor) {
        return runningJobs.stream().filter(j -> j.getDescriptor().equals(descriptor)).findFirst().orElse(null);
    }

}

File: src/main/kotlin/ro/sopa/lastfm/job/JobRegistry.kt
package ro.sopa.lastfm.job

import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

@Component
@Scope("singleton")
class JobRegistry {

    private val runningJobs: ArrayList<RunningJob> = ArrayList()

    fun register(descriptor: JobDescriptor) {

        var job = findRunningJob(descriptor)

        if(job != null) {
            throw RuntimeException("Job already running for $descriptor and is at $job.step/$job.total")
        }

        job = RunningJob(descriptor)

        runningJobs.add(job)

    }

    fun setTotalStep(descriptor: JobDescriptor, total: Int) {
        val runningJob = findRunningJob(descriptor) ?: throw RuntimeException("Invalid job: $descriptor")
        runningJob.total = total
    }

    fun setStep(descriptor: JobDescriptor, step: Int) {
        val runningJob = findRunningJob(descriptor) ?: throw RuntimeException("Invalid job: $descriptor")
        runningJob.step = step
    }

    fun finish(descriptor: JobDescriptor) {
        val runningJob = findRunningJob(descriptor) ?: throw RuntimeException("Invalid job: $descriptor")
        runningJobs.remove(runningJob)
    }

    fun findRunningJob(descriptor: JobDescriptor) : RunningJob? =
        runningJobs.firstOrNull { j -> j.descriptor == descriptor }

}